cdap-app-fabric/src/main/java/co/cask/cdap/metadata/LineageAdmin.java

```diff
@@ -217,8 +259,80 @@ private Lineage doComputeLineage(final NamespacedEntityId sourceData, long start
       }
     }
 
-    Lineage lineage = new Lineage(
-      Iterables.concat(Maps.transformValues(relations.asMap(), COLLAPSE_UNKNOWN_TYPE_FUNCTION).values()));
+    Lineage lineage;
+    if (rollup.contains("workflow")) {
+      // Get workflow Run IDs for all the programs in the relations
+      final Set<String> workflowIDs = new HashSet<>();
+      for (RelationKey key : relations.keySet()) {
+        Collection<Relation> allRelations = relations.get(key);
+        for (Relation relation : allRelations) {
+          ProgramId program = relation.getProgram();
+          RunRecordMeta runRecord = store.getRun(program, relation.getRun().getId());
+          if (runRecord != null && runRecord.getProperties().containsKey("workflowrunid")) {
+            String workflowRunId = runRecord.getProperties().get("workflowrunid");
+            workflowIDs.add(workflowRunId);
+          }
+        }
+      }
+
+      // Get Program IDs for workflow Run IDs
+      Map<ProgramRunId, RunRecordMeta> workflowRunRecordMap =
+        store.getRuns(ProgramRunStatus.ALL,
```

Comment:
This call potentially can be quite slow as it scans all run records. Since we already has the same of workflow ids we are interested, we should be able to pass that information the store so it can be fetch more efficiently, by constructing a much tighter sets of start/end row key to scan.